src/modules/Game.class.js

```diff
@@ -1,68 +1,271 @@
+      [0, 0, 0, 0],
+      [0, 0, 0, 0],
+      [0, 0, 0, 0],
+      [0, 0, 0, 0],
+    ],
+  ) {
+    this.initialState = initialState;
+    this.state = this.initialState.map((row) => [...row]);
+    this.status = Game.Status.idle;
+    this.score = 0;
+  }
+
+  moveLeft() {
+    if (this.getStatus() !== Game.Status.playing) {
+      return;
+    }
+
+    const newState = this.state.map((row) => this.move(row));
+
+    if (JSON.stringify(this.getState()) !== JSON.stringify(newState)) {
+      this.updateGameState(newState);
+      this.insertTiles();
+    }
+  }
+
+  moveRight() {
+    if (this.getStatus() !== Game.Status.playing) {
+      return;
+    }
+
+    const reversedState = this.state.map((row) => [...row].reverse());
+
+    const newState = reversedState.map((row) => {
+      return this.move(row).reverse();
+    });
+
+    if (JSON.stringify(this.getState()) !== JSON.stringify(newState)) {
+      this.updateGameState(newState);
+      this.insertTiles();
+    }
+  }
+
+  moveUp() {
+    if (this.getStatus() !== Game.Status.playing) {
+      return;
+    }
+
+    const rotateState = this.rotateLeft(this.getState());
+
+    const newState = rotateState.map((row) => this.move(row));
+
+    const unRotateState = this.rotateRight(newState);
+
+    if (JSON.stringify(this.getState()) !== JSON.stringify(unRotateState)) {
+      this.updateGameState(unRotateState);
+      this.insertTiles();
+    }
+  }
+
+  moveDown() {
+    if (this.getStatus() !== Game.Status.playing) {
+      return;
+    }
+
+    const rotateState = this.rotateRight(this.getState());
+
+    const newState = rotateState.map((row) => this.move(row));
+
+    const unRotateState = this.rotateLeft(newState);
+
+    if (JSON.stringify(this.getState()) !== JSON.stringify(unRotateState)) {
+      this.updateGameState(unRotateState);
+      this.insertTiles();
+    }
+  }
+
+  move(row) {
+    const newRow = [];
+    let i = 0;
+
+    while (i < row.length) {
+      const current = row[i];
+
+      if (current) {
+        let merged = false;
+
+        for (let j = i + 1; j < row.length; j++) {
+          const next = row[j];
+
+          if (next === current) {
+            newRow.push(current * 2);
+            this.updateGameScore(current * 2);
+            merged = true;
+            i = j + 1;
+            break;
+          } else if (next) {
+            newRow.push(current);
+            merged = true;
+            i = j;
+            break;
+          }
+        }
+
+        if (!merged) {
+          newRow.push(current);
+          i++;
+        }
+      } else {
+        i++;
+      }
+    }
+
+    return [...newRow, ...Array(row.length - newRow.length).fill(0)];
+  }
+
+  rotateRight(matrix) {
+    const n = matrix.length;
+    const rotatedMatrix = [];
+
+    for (let i = 0; i < n; i++) {
+      rotatedMatrix.push([]);
+
+      for (let j = 0; j < n; j++) {
+        rotatedMatrix[i].unshift(matrix[j][i]);
+      }
+    }
+
+    return rotatedMatrix;
```

Comment:
The `rotateRight` and `rotateLeft` methods could be optimized by not using `unshift` and `push` inside a loop, which can be costly operations due to array reindexing.